### DISCUSSION TITLE
Tab width and multi-row

### DISCUSSION THREAD

**Discussion by beefjerky9** [onemen/TabMixPlus#285] [General]
So, I'm not sure if this would be considered an issue or not, so I'll start here.

I have multi-row enabled with a min/max tab width of 150/175. It seems that sometimes the bottom row, if not full, will have the tabs be wider than the upper rows. It does not always do this; for example, if my current window is maximized, it's fine, but if I mak the window a slightly smaller size, the bottom row tabs are wider. I can "fix" this by setting the min/max the same, but that's not really a good option for me, seeing as I sometimes use my laptop screen, and other times a docking station. I then end up with large spaces on the tab bar at times that are wasteful.

Is this something that's fixable, or is this intrinsic to the "newer, better" Firefox that we're stuck with?

Thanks!

EDIT: I wanted to add that I'm running Firefox v124.0.1

**Comment by pyrates999** [comment]
Considering this would need to be applied using rules:

1. what scenarios do you expect the tabs width to be 150
2. what scenarios do you expect tabs width to be 175?
3. what scenarios do you expect tabs width to be between 150 and 175?

**Reply by beefjerky9** [reply]
Hmm, perhaps I should clarify my expectations. It isn't about a particular width, as much as the inconsistent width between the rows. 

For example, when this happens, rows 1 and 2 might show with a width of 150, while the tabs in row 3 end up with a width of 175. I don't care which they end up at, as much as the 3 rows matching, and not having a ton of wasted space on the side.

I realize it's not the end of the world, or a large functional problem. However, it is distracting, and can sometimes messes me up when looking through the tabs. It did not do this in the "old" Tab Mix Plus with the old Firefox. So, my question was more of whether this "issue" is something that is due to the way the newer Firefox versions work, or whether it something can be "fixed" in Tab Mix Plus.

Thanks!

**Reply by pyrates999** [reply]
Ok, how did the old firefox behave then?  I'm assuming it is version 56 since that is the last version that supported the old tab mix plus extension.

**Reply by beefjerky9** [reply]
The old Firefox simply kept all the tabs in all the rows the same width. That width could change based on window size and number of tabs, but all the rows, even if the last row isn't full, were consistent. I'm honestly not sure how else to explain the difference.

**Reply by pyrates999** [reply]
So it would try to fit all tabs in the number of specified rows, adjusting the width of all tabs based on the minimum and maximum width settings, so it would fit.

Lets say you had 3 rows.  Each row was 100 pixels.  The minimum tab length is 30, while the maximum tab length is 70.

With 6 tabs open, using the above settings, you would expect each tab to be 50 pixels, with each row showing 2 tabs.

But where it gets complicated is what do you expect to happen when you have 7 tabs open.  Should rows 1 and 2 tabs be 50 pixels each, showing 2 tabs on rows 1 and 2?  While row 3 tabs are 33 pixels each showing the last tab?

Or should row 1 and 2 have 3 tabs each with each one 33 pixels long?  And row 3 have one tab with the tab 70 pixels long?

**Reply by beefjerky9** [reply]
Okay, clearly words aren't describing my issue, so let's try screenshots.

Here's the "issue" I'm having:
![DiffWidth](https://github.com/onemen/TabMixPlus/assets/40005415/73c83ca4-2a6c-4567-a1a3-acb5d29f95df)

I don't like how the tabs on the bottom row are a different width than the two other rows.  

---

Here's what happened after un-maximizing and maximizing the window. This is also how it consistently behaved in the old Tab Mix Plus/Firefox:
![SameWidth](https://github.com/onemen/TabMixPlus/assets/40005415/9bbdaeb2-3e17-43d5-9882-e236860d9a26)

All tabs are the same width. I don't care if there are less tabs on one of the rows, I just want them the same width as the other rows.

---

Here's what I'm talking about when I say "wasted space:"
![WastedSpace](https://github.com/onemen/TabMixPlus/assets/40005415/839a8ca5-a476-4374-84d3-7582f2b8d940)

This is, of course, what happens if I set a fixed width rather than a variable width, and the window isn't just the right size. This is not a desirable outcome, both aesthetically and functionally.

Hopefully things are clearer now. Thanks.

**Reply by pyrates999** [reply]
One thing I noticed is that in the first screen shot, the new tab button is to the right of the last tab.  Where as the second and third screen shot, the new tab button is on the side where the tab list button is.

So when the new tab button is beside the last tab, when un-maximizing and maximizing the window, it gets moved to the side.

So it might just be the new tab button is not staying where you expect it to be.  Do you have it set where the new tab button is beside the last tab or on the side?

If it's expecting to be beside the last tab button, but it is on the side instead, that may cause a miscalculation of what length to use for the tabs shown per row.  So it finds it can't fit the tabs in that row, and shrinks them maybe to the minimum.  And thus you get extra space instead.  Same thing happens in the second row here.

So this is all caused by that new tab button not staying beside the last tab, and instead moving itself to the side of all the tabs.

Does this make sense?

**Reply by beefjerky9** [reply]
I noticed that too, and changed the option for that button from "after last tab," to "on right side." This takes care of the moving button issue, and it now stays next to the other buttons. However, the tabs in the last row still do the same thing.

**Reply by pyrates999** [reply]
here's what I mean:

![image](https://github.com/onemen/TabMixPlus/assets/106135298/ed12d2b4-dd61-4626-a81a-540abdd7d1a8)

**Reply by pyrates999** [reply]
so I think this is caused by that moving new tab button, and thus something for the developer to fix here.

**Reply by beefjerky9** [reply]
Yes, I know, and that's what my reply was referring to. I changed the option for that button to "on the right side" and now it stays put with the other 3 buttons. The last row of tabs still do the same thing, even after that has changed.

**Reply by pyrates999** [reply]
is the last row using the width of what the other rows are using?

**Reply by beefjerky9** [reply]
>is the last row using the width of what the other rows are using?

No, it is still doing the same thing, even with that the button option changed.

**Reply by pyrates999** [reply]
Please wait for a developer to respond then.

**Comment by onemen** [comment]
Since Firefox 113 all elements are flex boxes. In [#215](https://github.com/onemen/TabMixPlus/issues/215#issuecomment-1507925398) I tried to find the best solution to this issue.

If you want to shrink all tab to the minimum width try to add this css rule to `userChrome.css` file in your profile `chrome` folder

```css
#tabbrowser-tabs[orient="horizontal"]:not([widthFitTitle]) > #tabbrowser-arrowscrollbox[flowing=multibar] > .tabbrowser-tab:not([pinned]):not([tiletabs-single="hidden"]) {
  flex: 0 1 0% !important;
}
```

to enable `userChrome.css` set in about:config `toolkit.legacyUserProfileCustomizations.stylesheets` to `true`

**Reply by beefjerky9** [reply]
Thanks for the reply, as I did not see that issue. I would like to ask a followup question about the CSS code you posted here.

Does that code simply make all tabs be whatever the minimum width is set to in TMP? Or does it continue to adjust the width of the tabs, but force the last row to be same width as the previous two rows?

Thanks!

**Reply by onemen** [reply]
The css will set all tabs width, when in multi-rows to the minimum width
